### PR TITLE
Add support for displaying progress bar for package larger than 50k

### PR DESCRIPTION
**Pull request by markjeee**
Initial prototype for review (no tests yet).

### CODE REVIEW

**Comment by markjeee** [comment]
[Explore: progress indicator for uploads](https://trello.com/c/lMd1Droj/20-explore-progress-indicator-for-uploads)

File: lib/gemfury/command/app.rb [hunk 1]
```diff
@@ -295,6 +315,65 @@ def push_files(command, gem_paths)
...
+      end
+
+      buf
+    end
+
+    def method_missing(*args)
+      @io.send(*args)
+    end
+
+    def respond_to?(meth, include_all = false)
```
Comment (rykov) [inline lib/gemfury/command/app.rb]: Can you do `UploadIO` this as a `SimpleDelegator` to handle `respond_to?` and `method_missing`? Also, maybe `ProgressIO` better reflects the name?

File: lib/gemfury/command/app.rb [hunk 2]
```diff
@@ -257,6 +260,10 @@ def push_files(command, gem_paths)
       g.is_a?(String) ? File.new(g) : g rescue nil
     end.compact
 
+    if !options[:silent]
```
Comment (rykov) [inline lib/gemfury/command/app.rb]: Is there a way to check here that output stream supports TTY?

File: spec/gemfiles/Gemfile.18 [hunk 3]
```diff
@@ -21,6 +21,7 @@ group :development do
   gem 'highline', '< 1.7.0.pre'
   gem 'faraday', '< 0.10.0.pre'
   gem 'json', '< 2.0.0.pre'
+  gem 'progress_bar', '< 1.0.4'
```
Comment (rykov) [inline spec/gemfiles/Gemfile.18]: Looks like leftovers from previous implementation

File: lib/gemfury/command/app.rb [hunk 4]
```diff
@@ -295,6 +315,59 @@ def push_files(command, gem_paths)
     end
   end
 
+  C50K = 50000
+
+  class ProgressIO < SimpleDelegator
+    attr_reader :content_type, :original_filename, :local_path, :io
```
Comment (rykov) [inline lib/gemfury/command/app.rb]: You can probably get rid of `@io` and `attr_reader :io`, and use SimpleDelegator's `__getob__` instead. You could use `alias_method :io, :__getobj__`, if you want to keep the external API here.

Comment (markjeee) [reply lib/gemfury/command/app.rb]: Ahh good catch. I don't think i need this externally exposed. So will remove instead.

File: lib/gemfury/command/app.rb [hunk 5]
```diff
@@ -295,6 +315,59 @@ def push_files(command, gem_paths)
...
+
+      super(@io)
+    end
+
+    def show_bar?
+      @bar != nil
+    end
+
+    def read(length)
+      buf = @io.read(length)
```
Comment (rykov) [inline lib/gemfury/command/app.rb]:
```suggestion
      buf = __getobj__.read(length)
```

File: lib/gemfury/command/app.rb [hunk 6]
```diff
@@ -295,6 +315,59 @@ def push_files(command, gem_paths)
...
+      end
+
+      if filesize > C50K
+        title = 'Uploading %s ' % File.basename(fname)
+        @bar = ProgressBar.create(:title => title, :total => filesize)
+      else
+        @bar = nil
+      end
+
+      super(@io)
```
Comment (rykov) [inline lib/gemfury/command/app.rb]:
```suggestion
      super(io)
```

File: lib/gemfury/command/app.rb [hunk 7]
```diff
@@ -295,6 +315,59 @@ def push_files(command, gem_paths)
...
+        io = File.open(filename_or_io)
+        local_path = filename_or_io
+      end
+
+      fname ||= local_path
+
+      @content_type = content_type
+      @original_filename = File.basename(fname)
+      @local_path = local_path
+      @io = io
```
Comment (rykov) [inline lib/gemfury/command/app.rb]:
```suggestion

      if io.respond_to? :size
        filesize = io.size
      else
        filesize = io.stat.size
      end
```

**Pull request merged**